File: Frontend/trackr-administrador/src/app/shared/http/seguridad/acceso-ayuda.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { AccesoAyuda } from '@models/seguridad/acceso-ayuda';
import { AccesoAyudaSeccionado } from '@models/seguridad/acceso-ayuda-seccionado';

@Injectable(
  {
    providedIn: 'root'
  }
)
export class AccesoAyudaService {
  private dataUrl = 'accesoAyuda/';

  constructor(public http: HttpClient) {}

  public consultar(idAccesoAyuda: number): Observable<AccesoAyuda> {
    return this.http.get<AccesoAyuda>(this.dataUrl + `consultar/${idAccesoAyuda}`);
  }
  public consultarPorAcceso(idAcceso: number) {
    return this.http.get<AccesoAyuda[]>(this.dataUrl + `consultarPorAcceso/${idAcceso}`);
  }
  public consultarPorAccesoPorSeccion(idAcceso: number) {
    return this.http.get<AccesoAyudaSeccionado[]>(this.dataUrl + `consultarPorAccesoPorSeccion/${idAcceso}`);
  }

  public agregar(accesoAyuda: AccesoAyuda): Observable<void> {
    return this.http.post<void>(this.dataUrl + 'agregar', accesoAyuda);
  }

  public editar(accesoAyuda: AccesoAyuda): Observable<void> {
    return this.http.put<void>(this.dataUrl + 'editar', accesoAyuda);
  }

  public eliminar(idAccesoAyuda: number): Observable<void> {
    return this.http.delete<void>(this.dataUrl + `eliminar/${idAccesoAyuda}`);
  }

}

File: Backend/TrackrAPI/TrackrAPI/Services/Seguridad/AccesoAyudaService.cs
using TrackrAPI.Dtos.Seguridad;
using TrackrAPI.Models;
using TrackrAPI.Repositorys.Seguridad;

namespace TrackrAPI.Services.Seguridad
{
    public class AccesoAyudaService
    {
        private IAccesoAyudaRepository accesoAyudaRepository;
        private AccesoAyudaValidatorService accesoAyudaValidatorService;

        public AccesoAyudaService(IAccesoAyudaRepository accesoAyudaRepository,
            AccesoAyudaValidatorService accesoAyudaValidatorService)
        {
            this.accesoAyudaRepository = accesoAyudaRepository;
            this.accesoAyudaValidatorService = accesoAyudaValidatorService;
        }

        public AccesoAyuda Consultar(int idAccesoAyuda)
        {
            return accesoAyudaRepository.Consultar(idAccesoAyuda);
        }
        public IEnumerable<AccesoAyudaDto> ConsultarPorAcceso(int idAcceso)
        {
            return accesoAyudaRepository.ConsultarPorAcceso(idAcceso);
        }

        public AccesoAyudaDto ConsultarDto(int idAccesoAyuda)
        {
            return accesoAyudaRepository.ConsultarDto(idAccesoAyuda);
        }
        public IEnumerable<AccesoAyudaSeccionDto> ConsultarPorAccesoPorSeccion(int idAcceso)
        {
            IEnumerable<AccesoAyudaDto> accesoAyudas;
            accesoAyudas = accesoAyudaRepository.ConsultarPorAcceso(idAcceso);

            IEnumerable<int?> idSecciones;
            idSecciones = accesoAyudas.Select(a => a.IdAyudaSeccion).Distinct();

            List<AccesoAyudaSeccionDto> secciones = new();
            foreach (int? idSeccion in idSecciones)
            {
                IEnumerable<AccesoAyudaDto> accesosSeccion = accesoAyudas.Where(a => a.IdAyudaSeccion == idSeccion);

                AccesoAyudaSeccionDto seccion = new();
                seccion.IdAyudaSeccion = idSeccion;
                if (idSeccion == null)
                {
                    seccion.NombreAyudaSeccion = "Demás ayudas";
                }
                else
                {
                    seccion.NombreAyudaSeccion = accesosSeccion.FirstOrDefault().NombreAyudaSeccion;
                }
                seccion.AccesoAyudas = accesosSeccion;
                secciones.Add(seccion);
            }
            return secciones.OrderBy(a => a.IdAyudaSeccion == null).ThenBy(a => a.IdAyudaSeccion);
        }

        public AccesoAyuda Agregar(AccesoAyuda accesoAyuda)
        {
            accesoAyudaValidatorService.ValidarAgregar(accesoAyuda);
            accesoAyudaRepository.Agregar(accesoAyuda);
            return accesoAyuda;
        }

        public void Editar(AccesoAyuda accesoAyuda)
        {
            accesoAyudaValidatorService.ValidarEditar(accesoAyuda);
            accesoAyudaRepository.Editar(accesoAyuda);
        }

        public void Eliminar(int idAccesoAyuda)
        {
            AccesoAyuda accesoAyuda = accesoAyudaRepository.Consultar(idAccesoAyuda);
            accesoAyudaValidatorService.ValidarEliminar(idAccesoAyuda);
            accesoAyudaRepository.Eliminar(accesoAyuda);
        }

    }
}
